src/hooks.h

```diff
@@ -0,0 +1,256 @@
+__NL__                                                                         \
+__NL__         MAP(INVOKE_HOOK_FOR_PLUGIN, __VA_ARGS__)                        \
+__NL__                                                                         \
+__NL__         return hook_return_val;                                         \
+__NL__      }                                                                  \
+__NL__                                                                         \
+__NL__      /* An overloaded empty arguments version of apply to support       \
+__NL__         hooks with void arguments */                                    \
+__NL__      template<typename Hook__,                                          \
+__NL__                     /* Invokes the hook method on the plugin */         \
+__NL__               typename... Args__ /* The hook method call arguments */   \
+__NL__      >                                                                  \
+__NL__      static auto apply() -> typename Hook__::ReturnValue {              \
+__NL__         MAP(INVOKE_EMPTY_ARGS_HOOK_FOR_PLUGIN, __VA_ARGS__)             \
+__NL__      }                                                                  \
+__NL__   };                                                                    \
+__NL__                                                                         \
+__NL__   PluginHookAdapter<HookLoop> hookLoop;
+
+// This macro is supposed to be called at the end of the begin() method
+// of the firmware sketch to connect the plugin hooks.
+//
+// Unfortunately, the c++ standard (until c++17) does not allow for
+// local class templates, neither for classes that feature template functions.
+// Thus we have to separate initialization and connection of plugins.
+// Otherwise, everything could be called from within the begin() method
+// of the firmware sketch.
+//
+#define KALEIDOSCOPE_CONNECT_PLUGINS                                           \
+   Kaleidoscope.connectPlugins(&hookLoop);
+
+// The following invoke macros are meant to be used with
+// KALEIDOSCOPE_CONNECT_PLUGINS in conjunction with the MAP macro
+// that casts a specific operation on every member of
+// a variadic macro argument list (if present).
+//
+#define INVOKE_HOOK_FOR_PLUGIN(PLUGIN)                                         \
+__NN__                                                                         \
+__NN__   hook_return_val = Hook__::invoke(PLUGIN, hook_args...);               \
+__NL__                                                                         \
+__NL__   if (!ContinuationPredicate__::eval(hook_return_val)) {                \
+__NL__      return hook_return_val;                                            \
+__NL__   }
+
+#define INVOKE_EMPTY_ARGS_HOOK_FOR_PLUGIN(PLUGIN)                              \
+                                                                               \
```

Comment:
Why the extra blank line here? There isn't one for `KALEIDOSCOPE_CONNECT_PLUGINS` above, and that's easier to read. If it's because you want to consistently use an extra blank line for macro functions, then prepending the `__NL__` should also be done consistently.